lua/diffview/vcs/utils.lua

```diff
@@ -0,0 +1,602 @@
+
+  local namestat_job = Job:new({
+    command = adapter:bin(),
+    args = adapter:get_namestat_args(args),
+    cwd = adapter.ctx.toplevel,
+    on_exit = on_exit
+  })
+  local numstat_job = Job:new({
+    command = adapter:bin(),
+    args = adapter:get_numstat_args(args),
+    cwd = adapter.ctx.toplevel,
+    on_exit = on_exit
+  })
+
+  namestat_job:start()
+  numstat_job:start()
+  latch:await()
+  local out_status
+  if not (#namestat_job:result() == #numstat_job:result()) then
+    out_status = ensure_output(2, { namestat_job, numstat_job }, "git.utils>tracked_files()")
+  end
+
+  if out_status == JobStatus.ERROR or not (namestat_job.code == 0 and numstat_job.code == 0) then
+    callback(utils.vec_join(namestat_job:stderr_result(), numstat_job:stderr_result()), nil)
+    return
+  end
+
+  local numstat_out = numstat_job:result()
+  local data = {}
+  local conflict_map = {}
+
+  for i, s in ipairs(namestat_job:result()) do
+    local status = s:sub(1, 1):gsub("%s", " ")
+    local name = s:match("[%a%s][^%s]*\t(.*)")
+    local oldname
+
+    if name:match("\t") ~= nil then
+      oldname = name:match("(.*)\t")
+      name = name:gsub("^.*\t", "")
+    end
+
+    local stats = {
+      additions = tonumber(numstat_out[i]:match("^%d+")),
+      deletions = tonumber(numstat_out[i]:match("^%d+%s+(%d+)")),
+    }
+
+    if not stats.additions or not stats.deletions then
+      stats = nil
+    end
+
+    if not (status == "U" and kind == "staged") then
+      table.insert(data, {
+        status = status,
+        name = name,
+        oldname = oldname,
+        stats = stats,
+      })
+    end
+
+    if status == "U" then
+      conflict_map[name] = data[#data]
+    end
+  end
+
+  if kind == "working" and next(conflict_map) then
+    data = vim.tbl_filter(function(v)
+      return not conflict_map[v.name]
+    end, data)
+
+    for _, v in pairs(conflict_map) do
+      table.insert(conflicts, FileEntry.with_layout(opt.merge_layout, {
+        adapter = adapter,
+        path = v.name,
+        oldpath = v.oldname,
+        status = "U",
+        kind = "conflicting",
+        rev_ours = adapter.Rev(RevType.STAGE, 2),
+        rev_main = adapter.Rev(RevType.LOCAL),
+        rev_theirs = adapter.Rev(RevType.STAGE, 3),
+        rev_base = adapter.Rev(RevType.STAGE, 1),
+      }))
+    end
+  end
+
+  for _, v in ipairs(data) do
+    table.insert(files, FileEntry.for_d2(opt.default_layout, {
+      adapter = adapter,
+      path = v.name,
+      oldpath = v.oldname,
+      status = v.status,
+      stats = v.stats,
+      kind = kind,
+      rev_a = left,
+      rev_b = right,
+    }))
+  end
+
+  callback(nil, files, conflicts)
+end, 7)
+
+---@param adapter VCSAdapter
+---@param left Rev
+---@param right Rev
+---@param opt git.utils.LayoutOpt
+---@param callback function
+local untracked_files = async.wrap(function(adapter, left, right, opt, callback)
+  Job:new({
+    command = adapter:bin(),
+    args = adapter:get_files_args(),
```

Comment:
`get_files_args`: if this method's parameter is optional, then it must be declared as such in it's annotations.